ibis/common/validators.py

```diff
@@ -89,11 +90,36 @@ def instance_of(klasses, arg, **kwargs):
     if not isinstance(arg, klasses):
         # TODO(kszucs): unify errors coming from various validators
         raise IbisTypeError(
-            f'Given argument with type {type(arg)} ' f'is not an instance of {klasses}'
+            f"Given argument with type {type(arg)} is not an instance of {klasses}"
         )
     return arg
 
 
+class lazy_instance_of(Validator):
+    """A version of `instance_of` that accepts class qualnames instead of "
+    "already imported classes.
+
+    Useful for delaying imports.
+    """
+
+    def __init__(self, classes):
+        classes = (classes,) if isinstance(classes, str) else tuple(classes)
+        self._classes = classes
+        self._check = lazy_singledispatch(lambda x: False)
```

Comment:
Would it be possible to factor out the lazy loading part from `lazy_singledispatch` without the dispatching functionality? 

```py
@validator
def lazy_instance_of(types, obj):
    concrete_types = lazy_load_types(types, module=get_module_name(obj))
    if not isinstance(obj, concrete_types):
        raise ...
    return obj
```

where `lazy_load_types` (with a better name) would correspond to the second step in the lazy singledispatch impl with a global lookup cache:

```py
            # 2. Check lazy implementations
             module = cls2.__module__.split(".", 1)[0]
             if lazy := lazy_lookup.get(module):
                 # Import all lazy implementations first before registering
                 # (which should never fail), to ensure an error anywhere
                 # doesn't result in a half-registered state.
                 new = {import_object(name): func for name, func in lazy.items()}
                 lookup.update(new)
                 # drop lazy implementations, idempotent for thread safety
                 lazy_lookup.pop(module, None)
```

No need to do this in this PR of course.

Reply:
If I understand what you're asking for correctly, that would only delay the imports until the first time this validator was hit, regardless of type used.

As written, this implementation only imports the lazily loaded types when an object from that lazy module is provided. If the user never imports shapely, then ibis won't either.

Reply:
No, I'd preserve the behaviour, so `lazy_instance_of("np.ndarray", [1, 2, 3])` would not load numpy either, hence the `lazy_load_types(types, module=get_module_name(obj))` module arg in my pseudocode above.